File: server/src/main/java/com/nexoffer/controller/ChatController.java
package com.nexoffer.controller;

import com.nexoffer.model.Profile;
import com.nexoffer.model.User;
import com.nexoffer.repository.ProfileRepository;
import com.nexoffer.service.LlmGatewayService;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/chat")
public class ChatController {

    private final ProfileRepository profileRepository;
    private final LlmGatewayService llmGatewayService;

    public ChatController(ProfileRepository profileRepository, LlmGatewayService llmGatewayService) {
        this.profileRepository = profileRepository;
        this.llmGatewayService = llmGatewayService;
    }

    @PostMapping
    public ResponseEntity<?> handleChat(@AuthenticationPrincipal User user, @RequestBody Map<String, Object> body) {
        String message = body.containsKey("message") ? body.get("message").toString() : null;

        if (message == null || message.trim().isEmpty()) {
            return ResponseEntity.badRequest().body(Map.of("success", false, "message", "Message cannot be empty."));
        }

        Optional<Profile> profileOpt = profileRepository.findByUserId(user.getId());
        String company = profileOpt.map(Profile::getCompany).orElse("Target Company");
        String role = profileOpt.map(Profile::getRole).orElse("Software Engineer");
        String resumeText = profileOpt.map(Profile::getResumeText).orElse("");
        String jobDescription = profileOpt.map(Profile::getJobDescription).orElse("");

        String systemPrompt = "You are the **NexOffer AI Career Assistant**, an intelligent, empathetic, and sharp technical interview mentor.\n" +
                "Target Company: " + company + "\n" +
                "Target Role: " + role + "\n" +
                "Stored Resume Summary: " + (resumeText.isEmpty() ? "Not yet uploaded" : "Available") + "\n" +
                "Stored JD: " + (jobDescription.isEmpty() ? "Not provided yet" : "Available") + "\n\n" +
                "Guidelines:\n" +
                "1. Provide concise, high-value advice and code snippets where requested.\n" +
                "2. Ground all answers in their target role and company.\n" +
                "3. Format cleanly with markdown (bolding, bullet points, code blocks).";

        StringBuilder formattedHistory = new StringBuilder();
        if (body.containsKey("conversationHistory") && body.get("conversationHistory") instanceof List) {
            List<?> history = (List<?>) body.get("conversationHistory");
            int start = Math.max(0, history.size() - 6);
            for (int i = start; i < history.size(); i++) {
                Object item = history.get(i);
                if (item instanceof Map) {
                    Map<?, ?> map = (Map<?, ?>) item;
                    String sender = "user".equals(map.get("sender")) ? "Candidate" : "NexOffer Coach";
                    formattedHistory.append(sender).append(": ").append(map.get("text")).append("\n");
                }
            }
        }

        String prompt = (formattedHistory.length() > 0 ? "Recent Conversation:\n" + formattedHistory + "\n\n" : "") + "Candidate Question: " + message;

        Map<String, String> context = new HashMap<>();
        context.put("company", company);
        context.put("role", role);
        if (!resumeText.isEmpty()) context.put("resumeText", resumeText);
        if (!jobDescription.isEmpty()) context.put("jobDescription", jobDescription);

        LlmGatewayService.AiResponse aiResponse = llmGatewayService.generateAIResponse(
                prompt,
                systemPrompt,
                context,
                false,
                "chat"
        );

        return ResponseEntity.ok(Map.of(
                "success", true,
                "reply", aiResponse.getContent(),
                "provider", aiResponse.getProvider()
        ));
    }
}

File: server/controllers/chatController.js
const { ProfileRepository } = require('../models/dbStore');
const llmGateway = require('../services/llmGateway');

// @desc    Chat with AI Career Assistant with full context injection
// @route   POST /api/chat
const handleChat = async (req, res, next) => {
  try {
    const { message, conversationHistory = [] } = req.body;

    if (!message || message.trim().length === 0) {
      return res.status(400).json({
        success: false,
        message: 'Message cannot be empty.',
      });
    }

    const profile = await ProfileRepository.findOne({ userId: req.user._id });
    const company = profile?.company || 'Target Company';
    const role = profile?.role || 'Software Engineer';
    const resumeText = profile?.resumeText || '';
    const jobDescription = profile?.jobDescription || '';

    const systemPrompt = `
You are the **NexOffer AI Career Assistant**, an intelligent, empathetic, and sharp technical interview mentor.
You have immediate access to the candidate's target profile:
- Target Company: ${company}
- Target Role: ${role}
- Stored Resume Summary: ${resumeText ? 'Available (Use it to give personalized answers)' : 'Not yet uploaded'}
- Stored Job Description: ${jobDescription ? 'Available (Reference specific requirements)' : 'Not provided yet'}

Guidelines:
1. Provide concise, direct, high-value advice and code snippets where requested.
2. Ground all answers in the context of their specific role (${role}) and company (${company}).
3. When asked to test or interview the candidate, ask ONE focused question at a time and evaluate their response.
4. Format output cleanly with markdown (bolding, bullet points, code blocks).
`;

    let formattedHistory = '';
    if (conversationHistory.length > 0) {
      formattedHistory = conversationHistory
        .slice(-6)
        .map((msg) => `${msg.sender === 'user' ? 'Candidate' : 'NexOffer Coach'}: ${msg.text}`)
        .join('\n');
    }

    const prompt = `${formattedHistory ? `Recent Conversation:\n${formattedHistory}\n\n` : ''}Candidate Question: ${message}`;

    const context = {
      company,
      role,
      resumeText,
      jobDescription,
    };

    const aiResult = await llmGateway.generateAIResponse({
      prompt,
      systemPrompt,
      context,
      jsonMode: false,
      moduleType: 'chat',
    });

    res.status(200).json({
      success: true,
      reply: aiResult.content,
      provider: aiResult.provider,
    });
  } catch (error) {
    next(error);
  }
};

module.exports = {
  handleChat,
};
